valhalla/common/telescope_states.py

```diff
@@ -15,116 +15,188 @@
+                event_type, event_reason = this_type, this_reason
+                break
+        return event_type, event_reason
+
+    def _lump_end(self, lump, next_event=None):
+        if not next_event or str(lump['telescope']) != str(self._telescope(next_event)):
+            return self.end
+        return self._es_to_datetime(next_event['timestamp'])
+
+    def _set_lump(self, event):
+        return {
+            'reasons': [event['_source']['reason']],
+            'types': [event['_source']['type']],
+            'start': self._es_to_datetime(event['_source']['timestamp']),
+            'telescope': self._telescope(event['_source']),
+            'latest_timestamp': self._es_to_datetime(event['_source']['timestamp'])
+        }
+
+    def _update_lump(self, lump, event):
+        if event['_source']['type'] not in lump['types'] or event['_source']['reason'] not in lump['reasons']:
+            lump['reasons'].append(event['_source']['reason'])
+            lump['types'].append(event['_source']['type'])
+        lump['latest_timestamp'] = self._es_to_datetime(event['_source']['timestamp'])
+        return lump
+
+    def get(self):
+        telescope_states = {}
+        current_lump = dict(reasons=None, types=None, start=None)
+
+        for event in self.event_data:
+            if self._telescope(event['_source']) not in self.available_telescopes:
+                continue
+
+            if current_lump['start'] is None:
+                current_lump = self._set_lump(event)
+                continue
+
+            if self._belongs_in_lump(event['_source'], current_lump):
+                current_lump = self._update_lump(current_lump, event)
+            else:
+                lump_end = self._lump_end(current_lump, event['_source'])
+                if lump_end < self.start:
+                    current_lump = self._set_lump(event)
+                    continue
+
+                telescope_states = self._update_states(telescope_states, current_lump, lump_end)
+                current_lump = self._set_lump(event)
+
+        if current_lump['start']:
+            lump_end = self._lump_end(current_lump)
+            telescope_states = self._update_states(telescope_states, current_lump, lump_end)
+
+        return telescope_states
+
+    def _update_states(self, states, lump, lump_end):
+        if lump['telescope'] not in states:
+            states[lump['telescope']] = []
+
+        event_type, event_reason = self._categorize(lump)
+        states[lump['telescope']].append(
+            {
+                'telescope': str(lump['telescope']),
+                'event_type': event_type,
+                'event_reason': event_reason,
+                'start': max(self.start, lump['start']),
+                'end': lump_end
+            }
+        )
+        return states
+
+    def _es_to_datetime(self, timestamp):
+        return datetime.strptime(timestamp, ES_STRING_FORMATTER).replace(tzinfo=timezone.utc)
```

Comment:
This seems like something that should be in a more generic location (utils or something)